compiler/passes/normalize.cpp

```diff
@@ -2833,18 +2833,76 @@ static void fixupExportedArrayFormals(FnSymbol* fn) {
         // TODO: handle call expression stuff.
       }
 
-      // Create a representation of the array argument that is accessible
-      // outside of Chapel (in the form of a pointer and a corresponding size)
-      formal->typeExpr->replace(new BlockStmt(new SymExpr(dtExternalArray->symbol)));
+      SymExpr* dom = toSymExpr(call->get(1));
+      if (dom != NULL && dom->symbol() == gNil) {
+        // The domain is nil.  Try to make a chpl_external_array with it.
+        // If that doesn't work, the user must be more explicit with their
+        // domain
+        formal->typeExpr->replace(new BlockStmt(new SymExpr(dtExternalArray->symbol)));
```

Comment:
I don't see the branch below using `dom`. Couldn't this branch be eliminated? Is the issue that we're trying to keep historical behavior that `export proc f(A:[] int)` assumes that A is default rectangular / external_array ? If so, I think it's worth a comment (about the user-facing behavior you're trying to create, rather than what the code here is doing).

Reply:
In the case where the domain is nil, we can't finish the chpl__buildArrayRuntimeType call to get the array type because those calls require having all their arguments.  We get around this for array arguments in normal functions by replacing the chpl__buildArrayRuntimeType call with the _array type and letting the actual determine the final type of the formal.  But since our actual is a type without that information visible on it, we don't have that recourse so have to fall back to chpl_external_array

Reply:
I don't think this is a fundamental problem, because there is at run-time a link from the array to its domain. This can be found from the array class instance. However the *compile time* type needs some strategy, just as varying the element type does.